runtime/include/MemRefUtils.hpp

```diff
@@ -0,0 +1,161 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstring>
+
+#include <Exception.hpp>
+
+extern "C" {
+void *_mlir_memref_to_llvm_alloc(size_t size);
+void *_mlir_memref_to_llvm_aligned_alloc(size_t alignment, size_t size);
+void _mlir_memref_to_llvm_free(void *ptr);
+}
+
+template <typename T, size_t R> struct MemRefT {
+    T *data_allocated;
+    T *data_aligned;
+    size_t offset;
+    size_t sizes[R];
+    size_t strides[R];
+};
+
+template <typename T, size_t R> class MemRefView {
+  private:
+    const MemRefT<T, R> *buffer;
+    size_t tsize; // total size
+
+  public:
+    class MemRefIter {
+      private:
+        const MemRefT<T, R> *buffer;
+        size_t loc; // physical index
+
+      public:
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = T;
+        using difference_type = std::ptrdiff_t;
+        using pointer = T *;
+        using reference = T &;
+
+        MemRefIter(const MemRefT<T, R> *_buffer, size_t _idx = 0) : buffer(_buffer)
+        {
+            static_assert(R == 1, "[Class: MemRefIter] Error in Catalyst Runtime: Non-Iterable "
```

Comment:
Do we not need iterators for other ranks?

Reply:
I've seen some implementations in `mlir_c_runner`, but generally, I don't understand the definition of iterators for R > 1 and we don't need them in the scope of our runtime.

Reply:
Doesn't sample return a 2D memref?

Looking at the current version of the sample method and the way it allocates a temporary vector and then clones it into the memref is pretty awkward, shouldn't we writing straight into the destination memref?
```py
    std::vector<double> samples_vec(li_samples.size());
    for (size_t shot = 0; shot < shots; shot++) {
        for (size_t wire = 0; wire < numQubits; wire++) {
            samples_vec[shot * numQubits + wire] =
                static_cast<double>(li_samples[shot * numQubits + wire]);
        }
    }
    MemRefT<double, 2> src{
        samples_vec.data(), samples_vec.data(), 0, {shots, numQubits}, {numQubits, 1}};
    samples.clone(src);
```

Reply:
> Doesn't sample return a 2D memref?

Yes, it obviously does. But my point is that iterators usually are referring to 1-dim sequence of data and to use iterators for multi-dimensional memrefs, you need iterators for every dimension and use them in a nested loop. I didn't implement this because of not seeing much of use cases for it in the runtime.

The other way around implementing this Iterator (which is not really my favourite approach) is by implementing an iterator that traverses through all the elements of a memref with some "pre-defined" ordering of dimensions. I can obviously implement this if you guys see this less weird than the current implementation of sample :)

Reply:
It looks like `mlir_c_runner` does not define an iterator for every rank. Just for rank 0 and rank N (templated over rank, so rank must be known at compile-time). It also provides iterators for unranked memrefs (these are not templated so knowing rank at compile time is not necessary), which thinking into the future, we do want to support. I do think that all of this is out of scope for this PR, but I just want to express my preference for having this dependency on the runtime as we already have this dependency on the generated function. Ideally, we would only build mlir_c_runner without building all mlir and cache. Which I think would address the concern of longer CI times for building the runtime.